### PR TITLE
Add syntax highlighting and LSP for Dockerfiles

### LINKED ISSUE #504: Docker & Docker Compose support
**Before you begin**
Check the backlog of issues to reduce the chances of creating duplicates; if an issue already exists, place a `+1` (👍) on it.

Checked, no duplicate

**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Dockerfile & docker-compose.yml files are extremely common in my work and having syntax support would be great! I would hope for the following features (in order of what I priority I personally would expect)

- Syntax checking for errors
- Code completion
- Code reformatting
- Above features depend on version of Dockerfile and docker-compose.yml file (which are contained inside the file itself)

References:
https://docs.docker.com/engine/reference/builder/
https://docs.docker.com/compose/compose-file/

**Describe alternatives you've considered**
Alternative would be to not have these languages available. This would be possible, but editors commonly support them (idea suite, vscode etc. etc.)

Dockerfile & docker-compose.yml are dialects of Yaml so I would expect the solution to depend on yaml support.

**Screenshots**
-

### LINKED ISSUE #5311: Docker & Docker Compose support
**Before you begin**
Check the backlog of issues to reduce the chances of creating duplicates; if an issue already exists, place a `+1` (👍) on it.

Checked, no duplicate

**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Dockerfile & docker-compose.yml files are extremely common in my work and having syntax support would be great! I would hope for the following features (in order of what I priority I personally would expect)

- Syntax checking for errors
- Code completion
- Code reformatting
- Above features depend on version of Dockerfile and docker-compose.yml file (which are contained inside the file itself)

References:
https://docs.docker.com/engine/reference/builder/
https://docs.docker.com/compose/compose-file/

**Describe alternatives you've considered**
Alternative would be to not have these languages available. This would be possible, but editors commonly support them (idea suite, vscode etc. etc.)

Dockerfile & docker-compose.yml are dialects of Yaml so I would expect the solution to depend on yaml support.

**Screenshots**
-

### PR DESCRIPTION
**Pull request by everettraven**
Looks like it could resolve zed-industries/extensions#504 ? I haven't messed with any docker-compose stuff but seeing as that is YAML based and there is already YAML support this should address the Dockerfile syntax highlighting and LSP support.

I'm also a Rust newbie so if there is anything that looks off please do let me know!

A screenshot showing syntax highlighting when running via `cargo run`:
Screenshot 2024-01-27 at 8 28 25 PM

Release Notes:

- Added syntax highlighting and LSP for Dockerfiles

### CODE REVIEW

**Comment by everettraven** [comment]
@cla-bot check

File: crates/zed/src/languages/dockerfile.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    ) -> Option<LanguageServerBinary> {
+        get_cached_server_binary(container_dir, &*self.node).await
+    }
+}
+
+async fn get_cached_server_binary(
+    container_dir: PathBuf,
+    node: &dyn NodeRuntime,
+) -> Option<LanguageServerBinary> {
+    (|| async move {
```
Comment (maxdeviant) [inline crates/zed/src/languages/dockerfile.rs:98]: We have an `async_maybe!` macro in our `util` crate that can make this a bit nicer than having the IIFE.

Here's an example of it used in another LSP adapter: https://github.com/zed-industries/zed/blob/5b9cc2619491c33b133254770ec2494010c2d016/crates/zed/src/languages/gleam.rs#L104

File: crates/zed/src/languages/dockerfile.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+impl DockerfileLspAdapter {
+    pub fn new(node: Arc<dyn NodeRuntime>) -> Self {
+        Self { node }
+    }
+}
+
+#[async_trait]
+impl LspAdapter for DockerfileLspAdapter {
+    fn name(&self) -> LanguageServerName {
+        LanguageServerName("dockerfile".into())
```
Comment (maxdeviant) [inline crates/zed/src/languages/dockerfile.rs:36]: For the full language server name I think we'll want to use the name of the "binary":

```suggestion
        LanguageServerName("docker-langserver".into())
```

**Comment by d1y** [comment]
Awesome, but I noticed that the syntax after PREFIX(run. etc) doesn't seem to be highlighted, I'm very impressed by the vscode plug-in, can you help me take a look: https://marketplace.visualstudio.com/items?itemName=jeff-hykin.better-dockerfile-syntax
image

**Comment by SomeoneToIgnore** [comment]
This looks close to be done, so it would be great to add a new docs entry to the https://github.com/zed-industries/zed/tree/main/docs/src/languages list before merging.

File: crates/zed/src/languages/dockerfile.rs (hunk 1 above)
Comment (everettraven) [reply crates/zed/src/languages/dockerfile.rs]: Oh nice, thanks for letting me know :) - done in the latest force push: https://github.com/zed-industries/zed/pull/6905/commits/e4e2b22da4754894adb384cfa4d7c7b52f6a28f2

**Comment by everettraven** [comment]
> Awesome, but I noticed that the syntax after PREFIX(run. etc) doesn't seem to be highlighted, I'm very impressed by the vscode plug-in, can you help me take a look: https://marketplace.visualstudio.com/items?itemName=jeff-hykin.better-dockerfile-syntax image

@d1y I updated the tree-sitter highlighting to get as close to this as I could. The new highlighting looks like this:
Screenshot 2024-01-29 at 9 32 04 AM

There is variable expansion highlighting where possible but there are some limitations as the tree-sitter-dockerfile crate doesn't currently have support for expansion in certain instructions (one is the `RUN` instruction) and no shell command syntax highlighting (it was either the whole shell command is highlighted or none, although this limitation could be due to me being more novice in messing with this tooling). That being said, I think this PR is "good enough" as is regarding the syntax highlighting and it could be incrementally improved in the future.

**Comment by everettraven** [comment]
> This looks close to be done, so it would be great to add a new docs entry to the https://github.com/zed-industries/zed/tree/main/docs/src/languages list before merging.

@SomeoneToIgnore Thanks for pointing that out! Rebased and added!

**Comment by everettraven** [comment]
@maxdeviant @SomeoneToIgnore Anything else needed to get this merged?

File: crates/zed/src/languages/dockerfile/config.toml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+name = "Dockerfile"
+path_suffixes = ["Dockerfile"]
+line_comments = ["# "]
+brackets = [
+    { start = "{", end = "}", close = true, newline = true },
```
Comment (SomeoneToIgnore) [inline crates/zed/src/languages/dockerfile/config.toml:5]: If we want syntax highlights for those, we need `brackets.scm`

**Review by SomeoneToIgnore — APPROVED** [review]
I looked at https://github.com/neondatabase/neon/blob/main/Dockerfile with this and looks solid, thank you for bringing this up.

One minor thing I see is this
image

I would expect ARG to be highlighted and both ARG and ENV values to be not highlighted.

I like what Intellij-based IDE do with bash code blocks and `--from` detection (e.g. recognizing `pg-build` and allowing to navigate to it)
image

but that's advanced stuff somebody can work later (presumably, LSP improvements are needed for the navigation part?)

**Comment by d1y** [comment]
Do you have any further questions about this PR? Can we merge? I really need this

**Comment by SomeoneToIgnore** [comment]
> Can we merge? I

Would be nice, but GH does not allow this for the PRs with conflicts. Neither can I push into the branch, so one small step has to be done by somebody.

https://github.com/zed-industries/zed/pull/6905#issuecomment-1921258868

image

**Comment by SomeoneToIgnore** [comment]
https://github.com/zed-industries/zed/pull/7977 made that happen thanks to @d1y , hence closing this stale one.

**Comment by everettraven** [comment]
@d1y Thanks for taking this over! I had some stuff come up and wasn't able to circle back around to address the comments/conflicts :(. Awesome to see this work get in!

**Pull request closed without merging**